internal/hcl/modules/loader.go

```diff
@@ -0,0 +1,217 @@
+// If not then it downloads the module from the registry or from a remote source and updates the module manifest with the latest metadata.
+func (m *ModuleLoader) Load() (*Manifest, error) {
+	var manifest *Manifest
+
+	_, err := os.Stat(m.manifestFilePath())
+	if errors.Is(err, os.ErrNotExist) {
+		log.Debugf("No existing module manifest file found")
+	} else if err != nil {
+		log.Debugf("Error checking for existing module manifest: %s", err)
+	} else {
+		manifest, err = readManifest(m.manifestFilePath())
+		if err != nil {
+			log.Debugf("Error reading module manifest: %s", err)
+		}
+	}
+
+	if manifest == nil {
+		manifest = &Manifest{}
+	}
+
+	m.cache.loadFromManifest(manifest)
+
+	metadatas, err := m.loadModules(m.Path, "")
+	if err != nil {
+		return nil, err
+	}
+
+	manifest.Modules = metadatas
+
+	err = writeManifest(manifest, m.manifestFilePath())
+	if err != nil {
+		log.Debugf("Error writing module manifest: %s", err)
+	}
+
+	return manifest, nil
+}
+
+// loadModules recursively loads the modules from the given path.
+func (m *ModuleLoader) loadModules(path string, prefix string) ([]*ManifestModule, error) {
+	manifestModules := make([]*ManifestModule, 0)
+
+	module, diags := tfconfig.LoadModule(path)
+	if diags.HasErrors() {
+		return nil, diags.Err()
+	}
+
+	for _, moduleCall := range module.ModuleCalls {
+		metadata, err := m.loadModule(moduleCall, path, prefix)
+		if err != nil {
+			return nil, err
+		}
+
+		manifestModules = append(manifestModules, metadata)
+
+		nestedManifestModules, err := m.loadModules(filepath.Join(m.Path, metadata.Dir), metadata.Key+".")
+		if err != nil {
+			return nil, err
+		}
+
+		manifestModules = append(manifestModules, nestedManifestModules...)
+	}
+
+	return manifestModules, nil
+}
+
+// loadModule loads the module metadata from the given module call.
+// It works by doing the following:
+// 1. Checks if the module is already downloaded and the version/source has not changed.
+// 2. Checks if the module is a local module.
+// 3. Checks if the module is a registry module and downloads it.
+// 4. Checks if the module is a remote module and downloads it.
+func (m *ModuleLoader) loadModule(moduleCall *tfconfig.ModuleCall, parentPath string, prefix string) (*ManifestModule, error) {
+	key := prefix + moduleCall.Name
+
+	manifestModule, err := m.cache.lookupModule(key, moduleCall)
+	if err == nil {
+		log.Debugf("Module %s already loaded", key)
+
+		// Test if we can actually load the module. If not, then we should try re-loading it.
+		// This can happen if the directory the module was downloaded to has been deleted and moved
+		// so the existing manifest.json is out-of-date.
+		_, diags := tfconfig.LoadModule(path.Join(m.Path, manifestModule.Dir))
+		if diags.HasErrors() {
+			log.Debugf("Module %s cannot be loaded, re-loading: %s", key, diags.Err())
+		} else {
+			return manifestModule, err
+		}
+	} else {
+		log.Debugf("Module %s needs loaded: %s", key, err.Error())
```

Comment:
```suggestion
		log.Debugf("Module %s needs loading: %s", key, err.Error())
```

Reply:
TIL that this is called "infinitival copula deletion" which is common in Scotland and some other places, but not elsewhere.